Feature

Dmitry92111 opened the pull request:
Первый вариант

Dmitry92111 committed:
Feature

Dmitry92111 committed:
Добавлен формат вводимых дат - ДД.ММ.ГГГГ, а также циклы проверки корректности введенных дат

Dmitry92111 committed:
Добавлены циклы проверки вводимых данных в условие капитализации процентов, в сумму вклада, а также в процентную ставку, чтобы программа не завершалась каждый раз при вводе некорректных данных

Sindicat commented on src/Main.java:
```diff
@@ -0,0 +1,156 @@
+        double amount = -1;
+        while (amount < 0) {
+            System.out.print("Введите сумму вклада в рублях РФ (разделитель между рублями и копейками - запятая): ");
+            String inputAmount = scanner.nextLine().replace(',', '.');
```
тогда логично сделать разделитель между рублями и копейками точку - и строчку с replace(',', '.') можно будет удалить совсем

Dmitry92111 replied:
Оставил пока что запятую, так как в РФ она вроде используется как десятичный разделитель (https://ru.wikipedia.org/wiki/%D0%94%D0%B5%D1%81%D1%8F%D1%82%D0%B8%D1%87%D0%BD%D1%8B%D0%B9_%D1%80%D0%B0%D0%B7%D0%B4%D0%B5%D0%BB%D0%B8%D1%82%D0%B5%D0%BB%D1%8C)

Sindicat commented on src/Main.java:
```diff
@@ -0,0 +1,156 @@
+        double interestRate = -1;
+        while (interestRate < 0) {
+            System.out.print("Введите процентную ставку (в % годовых без знака <%>, разделитель - запятая): ");
+            String inputInterestRate = scanner.nextLine().replace(',', '.');
```
и тут разделитель сразу точку сделать

Sindicat commented on src/Main.java:
```diff
@@ -0,0 +1,156 @@
+        /* Расчет без капитализации путем пересчета в счетчиках общего количества дней в високосных или невисокосных годах,
+        а также последующего вычисления процентов по формуле расчета процентов по вкладу*/
+
+        if (capitalization.equals("Нет")) {
```
вместо string переменной capitalization, лучше использовать boolean переменную isCapitlizationEnabled. Почему так? - у тебя в  переменной capitalization ожидается два значения всего с логическими ДА или НЕТ. striong переменную capitalization плохо исопльзовать, потому что эта переменная может хранить сколько угодно значений и читающий твой код (если он начнет читать код не с начала, а вот с этого момента) должен будет гадать: "А какие еще значения там лежат? Может ли там лежать значение "Не знаю", "Решу потом"  и т.д. 
а с переменной isCapitlizationEnabled - все очевидно, только два значения true(да), либо false(нет)

Sindicat replied:
P.S. заполнять переменную лучше isCapitlizationEnabled в начале, где у тебя объявляется  переменная capitalization

Sindicat commented on src/Main.java:
```diff
@@ -0,0 +1,156 @@
+import java.time.format.DateTimeParseException;
+
+public class Main {
+    public static void main(String[] args) {
```
Все в принципе норм для начала. Но в действительности никто код в одной функции не пишет, получается вот такая простыня и фиг разберешься потом где что происходит.
Разбей свою одну main функцию на несколько:
1) Функции которые считывают данные от пользователя и валидируют их
2) Функция расчета процентов с капитализацией 
3) Функция расчета процентов без капитализации
 и что-то мне подсказывает, что 2 и 3 функции можно будет объединить в одну

Пример:
```
DecimalFormat decimalFormat = new DecimalFormat("#.##");
DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy");

System.out.print("Происходит ли капитализация процентов по вкладу в последний календарный день месяца? (Введите Да или Нет): ");
String capitalization = scanner.nextLine();

while (!capitalization.equals("Нет") && !capitalization.equals("Да")) {
    System.out.println("Неизвестная команда в строке ввода");
    System.out.print("Происходит ли капитализация процентов по вкладу в последний календарный день месяца? (Введите Да или Нет): ");
    capitalization = scanner.nextLine();
}
```

вынеси в отдельную функцию и назови например
```
private static boolean readCapitalizationCondition(Scanner scanner) {
// тут твой код считывания с консоли и валидации
return capitalization.equals("Да")
}
```
в main функции будет тогда так:
var  isCapitalizationEnabled = readCapitalizationCondition(scanner)

Dmitry92111 committed:
Разбил все по методам + сделал условие о капитализации процентов с помощью переменной типа boolean

Dmitry92111 committed:
заменил условие о капитализации процентов с помощью переменной типа boolean на список enum с целью дальнейшей доработки капитализации в последний рабочий день месяца, заменил DecimalFormat на NumberFormat для красивого вывода денежных сумм, объединил 2 схожих метода ввода дат (startDate & endDate) в один, немного улучшил названия переменных

Dmitry92111 committed:
Заменил расчеты с помощью переменных double на BigDecimal

Dmitry92111 committed:
Небольшие косметические улучшения

Dmitry92111 committed:
Объединил методы по запросу у пользователя суммы вклада и процентной ставки + небольшие улучшения

Dmitry92111 committed:
Значительные изменения!
Добавил расчет процентов с капитализацией в последний рабочий день месяца, также исправил некорректно работающие формулы расчета процентов
